README.md
# Pandora

Pandora is a command-line utility for processing PDF files, powered by [Apache PDFBox].

## Requirements

* Java 11 or higher

## Basic Usage

```shell
pandora <command>
```

## Commands
### crop

Assigns a crop box to the PDF document.

```shell
pandora crop [--even] [--flip] [--odd] [--preserve-aspect] [-a=<numeric value or paper size name>] [-o=<output>] [--padding=<padding>] [--pages=<page|range(,page|range)*>] [-m=<top,right,bottom,left>, "bbox", or "text-bbox"]... <input>
```
#### \<input\>
Path to the original PDF document.

#### -a, --aspect=\<numeric value or paper size name\>
Page aspect ratio to be forced. e.g. `0.75`, `3:4`, `a4`

#### --even
Process only even pages.

#### --flip
Flip the margin, page by page.

#### -m, --margin=\<top,right,bottom,left\>, `bbox`, or `text-bbox`

Each margin can be specified in 1/72 inch or % unit.
Special value `bbox` means calculated bounding box of the page. Another value `text-bbox` means bounding box of the texts in the page. (default value: `bbox`)

#### -o, --output=\<output\>
Path to the converted PDF document.

#### --odd
Process only odd pages.

#### --padding=\<padding\>
Padding size in 1/72 inch added to bounding boxes.
(default value: `5`)

#### --pages=\<page|range(,page|range)*\>
Pages or page ranges delimited by comma.
Each range is in the form `<first page>-<last page>`,
both bounds are inclusive and one-indexed.
Page numbers can be prefixed with `b` when counted from the back cover.

#### --preserve-aspect
Preserve the original aspect ratio of pages.

#### Examples
##### **Cropping a PDF with the specified margin**

Cropping pages from the 2nd to the last with the magin of top, right, bottom specified in 1/72 inch. The omitted left margin will be the same as the right margin.

```shell
pandora crop --pages 2-b1 -m 50,40,30 -o Classic_Shell_Scripting.cropped.pdf Classic_Shell_Scripting.pdf
```
##### **Cropping a PDF with automatically calculated bounding boxes**

```shell
pandora crop "Zero Trust Networks.pdf"
```

### help
Displays help information about the specified command.

#### Examples
##### **Displaying help about a command**
```shell
pandora help crop
```

## Copyright Notice
Copyright 2020-2021 the original author or authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this product except in compliance with the License.
You may obtain a copy of the License at
<http://www.apache.org/licenses/LICENSE-2.0>

[Apache PDFBox]: https://pdfbox.apache.org


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### Changed
- Replaced the calculation of bounding boxes with more precise image-based algorithm. 
  The old algorithm can be activated by specifying `fast-bbox` value for `--margin` option.

## 0.4.0 - 2021-03-14
### Added
- `--aspect` option to specify the aspect ratio of the pages.
- `--padding` option to add padding to the calculated bounding boxes.

### Changed
- Page ranges now use `<first>-<last>` form.

### Fixed
- The calculation of bounding box now ignores the areas outside of the page.

## 0.3.0 - 2020-10-17
### Added
- `text-bbox` is now available as a `--margin` option argument, which will calculate the bounding box containing all texts in the page.
- `--flip` option which will flip the left and the right margin, page by page.
- Specifying multiple margins is now supported. When using two margins, the first margin will be applied to the odd pages and the second margin will be applied to the even pages.

### Changed
- Now `setAllSecurityToBeRemove()` method on `PDDocument` will be always called to remove security.

### Fixed
- A bug in parsing page range with the last page omitted.

## 0.2.0 - 2020-04-04
- First release.


src/main/java/org/leadpony/pandora/CropCommand.java
/*
 * Copyright 2020-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.leadpony.pandora;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.common.PDRectangle;

import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

/**
 * @author leadpony
 */
@Command(name = "crop", description = "Assigns crop box to the PDF")
class CropCommand extends AbstractCommand implements CroppingContext {

    @Option(names = { "-m", "--margin" },
            paramLabel = "<top,right,bottom,left>, \"bbox\", \"fast-bbox\", or \"text-bbox\"",
            description = {
                "Each margin can be specified in 1/72 inch or %% unit.",
                "Special value \"bbox\" calculates bounding box for each page.",
                "\"fast-bbox\" produces approximate bounding box for each page.",
                "\"text-bbox\" produces bounding box only from texts in each page.",
                "(default value: \"bbox\")"
            },
            defaultValue = "bbox")
    private List<Margin> margin;

    @Option(names = "--preserve-aspect",
            description = "Preserve the original aspect ratio of pages.")
    private boolean preserveAspect;

    @Option(names = "--flip",
            description = "Flip the margin, page by page.")
    private boolean flip;

    @Option(names = "--padding",
            description = {
                    "Padding size in 1/72 inch added to bounding boxes.",
                    "(default value: 5)"
            },
            defaultValue = "5")
    private int padding = 5;

    @Option(names = {"-a", "--aspect"},
            paramLabel = "<numeric value or paper size name>",
            description = {
                    "Page aspect ratio to be forced.",
                    "e.g. 0.75, \"3:4\", \"a4\""
            },
            converter = AspectConverter.class)
    private Float aspect;

    private CropStrategy strategy;

    @Override
    public int getPadding() {
        return padding;
    }

    @Override
    protected void beginProcessing(PDDocument doc) {
        List<Margin> margins = new ArrayList<>(this.margin);
        if (margins.size() == 1 && this.flip) {
            margins.add(margins.get(0).flip());
        }

        List<CropStrategy> strategies = margins.stream()
                .map(margin -> margin.createStrategy(doc, this))
                .collect(Collectors.toList());

        if (strategies.size() == 1) {
            this.strategy = strategies.get(0);
        } else {
            this.strategy = new FlippingCropStrategy(strategies);
        }
    }

    @Override
    protected void processPage(PDDocument doc, int pageIndex) {
        PDPage page = doc.getPage(pageIndex);
        resetCropBox(page);
        PDRectangle cropBox = strategy.getCropBox(doc, pageIndex);
        updateCropBox(page, cropBox);
    }

    private void resetCropBox(PDPage page) {
        page.setCropBox(page.getMediaBox());
    }

    private void updateCropBox(PDPage page, PDRectangle cropBox) {
        final PDRectangle mediaBox = page.getMediaBox();
        if (preserveAspect) {
            cropBox = adjustBoxAspect(cropBox, mediaBox);
        } else if (aspect != null) {
            cropBox = adjustBoxAspect(cropBox, aspect);
        }
        clipBox(cropBox, mediaBox);
        page.setCropBox(cropBox);
    }

    private static PDRectangle adjustBoxAspect(PDRectangle box, PDRectangle page) {
        float aspectRatio = page.getWidth() / page.getHeight();
        return adjustBoxAspect(box, aspectRatio);
    }

    private static PDRectangle adjustBoxAspect(PDRectangle box, float aspectRatio) {
        final float newAspectRatio = box.getWidth() / box.getHeight();
        if (newAspectRatio < aspectRatio) {
            float width = box.getHeight() * aspectRatio;
            float x = box.getLowerLeftX() - 0.5f * (width - box.getWidth());
            return new PDRectangle(x, box.getLowerLeftY(), width, box.getHeight());
        } else if (newAspectRatio > aspectRatio) {
            float height = box.getWidth() / aspectRatio;
            float y = box.getLowerLeftY() - 0.5f * (height - box.getHeight());
            return new PDRectangle(box.getLowerLeftX(), y, box.getWidth(), height);
        }
        return box;
    }

    private static void clipBox(PDRectangle box, PDRectangle mediaBox) {
        float minX = box.getLowerLeftX();
        float maxX = box.getUpperRightX();
        float minY = box.getLowerLeftY();
        float maxY = box.getUpperRightY();

        if (minX < mediaBox.getLowerLeftX()) {
            box.setLowerLeftX(mediaBox.getLowerLeftX());
        }
        if (maxX > mediaBox.getUpperRightX()) {
            box.setUpperRightX(mediaBox.getUpperRightX());
        }
        if (minY < mediaBox.getLowerLeftY()) {
            box.setLowerLeftY(mediaBox.getLowerLeftY());
        }
        if (maxY > mediaBox.getUpperRightY()) {
            box.setUpperRightY(mediaBox.getUpperRightY());
        }
    }
}
